File: src/middlewares/aggregator.ts
import type {
	DescribableInput,
	DescribableOutput,
	LanguageModelMiddleware,
	LanguageModelMiddlewareContext,
	LanguageModelMiddlewareNext,
} from "../describe.ts"

export const aggregator =
	<
		Input extends DescribableInput = DescribableInput,
		Output extends DescribableOutput = DescribableOutput,
	>(
		...middlewares: LanguageModelMiddleware<Input, Output>[]
	): LanguageModelMiddleware<Input, Output> =>
	async (
		context: LanguageModelMiddlewareContext<Input, Output>,
		next: LanguageModelMiddlewareNext<Input, Output>,
	) => {
		const chain = middlewares.reduceRight<
			LanguageModelMiddlewareNext<Input, Output>
		>(
			(
				prev: LanguageModelMiddlewareNext<Input, Output>,
				curr: LanguageModelMiddleware<Input, Output>,
			) =>
				async (
					ctx: LanguageModelMiddlewareContext<Input, Output>,
				): Promise<LanguageModelMiddlewareContext<Input, Output>> =>
					curr(ctx, prev),
			next,
		)

		return chain(context)
	}

File: go/middlewares/aggregator.go
package middlewares

import "github.com/nmnmcc/toolbox"

// Aggregator composes multiple middlewares into a single middleware.
func Aggregator[Input, Output any](middlewares ...toolbox.Middleware[Input, Output]) toolbox.Middleware[Input, Output] {
	return func(ctx toolbox.Context[Input, Output], next toolbox.Next[Input, Output]) (toolbox.CompletionContext[Input, Output], error) {
		// Build chain by reducing from right to left
		chain := next
		for i := len(middlewares) - 1; i >= 0; i-- {
			mw := middlewares[i]
			prev := chain
			chain = func(ctx toolbox.Context[Input, Output]) (toolbox.CompletionContext[Input, Output], error) {
				return mw(ctx, prev)
			}
		}
		return chain(ctx)
	}
}
